### DISCUSSION TITLE
Python as Import Script

### DISCUSSION THREAD

**Discussion by s3erjaeh** [gerbera/gerbera#1239] [Ideas]
Hey,

I got inspired by my Android phone which can group photos by location and detected faces. So I stumbled over gerbera with the great idea of scripting the import. Unfortunately I missing the functionality to include other JS modules and also face-clustering is more on the python side. So I played with gerbera to include python as second option to JS. It took me a while, but it’s now functional and you can play around on [my gerbera fork](https://github.com/s3erjaeh/gerbera). I converted the existing functionality of the JS Import script to python and add a working reverse-geocode section. If you like the idea I can provide you a pull request.

**Comment by KarlStraussberger** [comment]
Basically, I like the idea of having a second scripting interface.

Taking into account all the folks (including me) who implemented their own import layout, I see no way of replacing JavaScript altogether. Especially, I remember some compatibility issues between python2 and python3 and all the installation dependencies we need to fulfil.  That makes me a bit nervous.

My first suggestion is to add python support via conditional compilation and CMake-Option, supplying another type for virual-layout in config.xml. If we do so, we still need somebody who maintains the code, which might be a dead end for a second scripting interface.

**Reply by s3erjaeh** [reply]
I don't want to replace JS, but duktape (like its implemented here) has no support for including external modules. So cool stuff like reverse-geocode or face recognition are not possible. 

The basic setup I have in my code only depends on an header-only library (pybind11) which need to be included in cmake. Also Python3 should be installed on the build and target machine. All other dependencies like reverse-geocode are completely on side of the scripting folks and gerbera should not take liability about. 
I already included python as conditional CMake option and make it selectable in the config via the "virtual-layout: type" attribute. Also coexistence with JS is possible. 

If you check the python integration with pybind11 you can also see a way much smaller footprint in additional code in comparison with ducktape. All integration is done within [123 lines of code](https://github.com/s3erjaeh/gerbera/blob/master/src/content/scripting/py_runtime.cc)

I can understand your maintainer point, but I think the amount of the python-integration-code is manageable and the new possibilities are awesome.
